### PR TITLE
[FLINK-3226] Translation of scalar function substring()

### PR DESCRIPTION
**Pull request by twalthr**
This PR implements the scalar function `substring()` for the Table API on Calcite. Additionally, it already contains preparations for more built-in SQL functions. I implemented test utils for scalar functions `ScalarFunctionsTest` and added a similar concept to Calcites `CallImplementor` named `CallGenerator`.

For now, I kept the `Substring` expression node, but I would like to remove it and use a generic `Call` expression. I think we don't need a case class for every operator. What do you think?

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/CodeGenerator.scala [hunk 1]
```diff
@@ -702,12 +704,17 @@ class CodeGenerator(
         requireBoolean(operand)
         generateNot(nullCheck, operand)
 
+      // casting
       case CAST =>
         val operand = operands.head
         generateCast(nullCheck, operand, resultType)
 
-      case call@_ =>
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/CodeGenerator.scala]: Add the all case to the end to get a good error message.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/table/test/ScalarFunctionsTest.scala [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    val exprResult = ExpressionEvaluator.evaluate(testData, typeInfo, expr)
+    assertEquals(expected, exprResult)
+
+    val exprStringResult = ExpressionEvaluator.evaluate(
+      testData,
+      typeInfo,
+      ExpressionParser.parseExpression(exprString))
+    assertEquals(expected, exprStringResult)
+
+    // TODO test SQL expression
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/table/test/ScalarFunctionsTest.scala]: Resolve TODO?

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/table/test/StringExpressionsITCase.scala [hunk 3]
```diff
@@ -18,42 +18,20 @@
...
 import org.junit.runner.RunWith
 import org.junit.runners.Parameterized
+
 import scala.collection.JavaConverters._
-import org.apache.flink.api.table.codegen.CodeGenException
 
 @RunWith(classOf[Parameterized])
 class StringExpressionsITCase(mode: TestExecutionMode) extends MultipleProgramsTestBase(mode) {
 
-  @Test(expected = classOf[CodeGenException])
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/table/test/StringExpressionsITCase.scala]: Why did you remove these tests? The `ScalarFunctionsTest` does not test the feature end-to-end, right?

File: flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/table/test/StringExpressionsITCase.java [hunk 4]
```diff
@@ -40,46 +40,6 @@ public StringExpressionsITCase(TestExecutionMode mode) {
 		super(mode);
 	}
 
-	@Test(expected = CodeGenException.class)
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/table/test/StringExpressionsITCase.java]: Why did you remove these tests? The `ScalarFunctionsTest` does not test the feature end-to-end, right?

**Comment by fhueske** [comment]
I had only a few minor comments. Looks mostly good.

File: flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/table/test/StringExpressionsITCase.java (hunk 4 above)
Comment (twalthr) [inline flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/table/test/StringExpressionsITCase.java]: No, the `ScalarFunctionsTest` does not test end-to-end. I don't think this is necessary for scalar functions, because they don't affect optimization/plan translation. When we will have more functions (e.g. 40 of them), we would have 40 end-to-end tests just implementing a MapFunction and evaluating an expression. That's why I implemented the `ExpressionEvaluator`.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/table/test/ScalarFunctionsTest.scala (hunk 2 above)
Comment (twalthr) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/table/test/ScalarFunctionsTest.scala]: Once we have a SQL parser ready, I will resolve this TODO ;-)

File: flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/table/test/StringExpressionsITCase.java (hunk 4 above)
Comment (fhueske) [inline flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/table/test/StringExpressionsITCase.java]: Agreed, we should not have end-to-end tests for all scalar functions. But it would be good to tests at least some representative functions, IMO.

Comment (twalthr) [inline flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/table/test/StringExpressionsITCase.java]: Yes, you are right, testing at least one function end-to-end makes sense. I will add them again.

**Comment by twalthr** [comment]
@fhueske build succeeded

**Comment by twalthr** [comment]
If there are no objections, I would like to merge this later.

**Comment by fhueske** [comment]
Thanks for the update. Looks good to merge
